packages/components/src/internal/renderers.tsx

```diff
@@ -98,169 +115,303 @@ export const EditableColumnTitle: FC<EditableColumnTitleProps> = memo(props => {
-    selectable?: boolean;
 }
 
-// exported for jest testing
-export const HeaderCellDropdown: FC<HeaderCellDropdownProps> = memo(props => {
+interface HeaderCellDropdownMenuProps extends SharedHeaderCellProps {
+    allowColFilter: boolean;
+    allowColSort: boolean;
+    colFilters: Filter.IFilter[];
+    isSortAsc: boolean;
+    isSortDesc: boolean;
+    onEditTitleClicked?: () => void;
+    open: boolean;
+    queryColumn: QueryColumn;
+    setOpen: (open: boolean) => void;
+}
+
+export const HeaderCellDropdownMenu: FC<HeaderCellDropdownMenuProps> = memo(props => {
     const {
-        i,
-        column,
-        handleSort,
-        handleFilter,
+        allowColFilter,
+        allowColSort,
+        colFilters,
         handleAddColumn,
+        handleFilter,
         handleHideColumn,
-        headerClickCount,
+        handleSort,
+        isSortAsc,
+        isSortDesc,
         model,
-        onColumnTitleChange,
-        onColumnTitleEdit,
+        onEditTitleClicked,
+        open,
+        queryColumn,
+        setOpen,
     } = props;
-    const col: QueryColumn = column.raw;
-    const [open, setOpen] = useState<boolean>();
-    const [editingTitle, setEditingTitle] = useState<boolean>(false);
-    const wrapperEl = useRef<HTMLSpanElement>();
-    const view = useMemo(() => model?.queryInfo?.getView(model?.viewName, true), [model?.queryInfo, model?.viewName]);
-
-    const allowColSort = handleSort && col?.sortable;
-    const allowColFilter = handleFilter && col?.filterable;
-    const allowColumnViewChange = (handleHideColumn || handleAddColumn) && !!model;
-    const includeDropdown = allowColSort || allowColFilter || allowColumnViewChange;
-
-    useEffect(() => {
-        return () => {
+    const showGridCustomization = handleHideColumn || handleAddColumn;
+    const toggleEl = useRef<HTMLSpanElement>();
+    const menuEl = useRef<HTMLUListElement>();
+    const portalRef = usePortalRef('header-cell-dropdown-menu-portal');
+    // Note: We need to make sure we cancel all events in our menu handlers or we also trigger the click handler in
+    // HeaderCellDropdown, which will reset the open value to true, which will keep the menu open.
+    const openFilterPanel = useCallback(
+        (_: any, event: SyntheticEvent) => {
+            cancelEvent(event);
             setOpen(false);
-        };
-    }, []);
-
-    const onToggleClick = useCallback(
-        (shouldOpen: boolean, evt?: any) => {
-            if (!includeDropdown || editingTitle) return;
-
-            // when menu is closed skip any clicks on icons by just checking for span el type
-            if (shouldOpen && evt && evt.target.tagName.toLowerCase() !== 'span') return;
-
-            setOpen(shouldOpen);
+            handleFilter(queryColumn, false);
         },
-        [includeDropdown]
-    );
-
-    const _handleFilter = useCallback(
-        (remove?: boolean) => {
-            handleFilter(col, remove);
+        [queryColumn, handleFilter]
```

Comment:
should dependencies array here include `setOpen`?

Reply:
Yes.